ObjectMapperCustomizer with Hibernate5Module for lazyloading

**Describe the bug**
Jackson ObjectMapperCustomizer not working as expected. 

**Expected behavior**
If the  Hibernate5Module would be set with the Hibernate5Module.Feature.SERIALIZE_IDENTIFIER_FOR_LAZY_NOT_LOADED_OBJECTS feature, i would expect that on every jax-rs endpoint with the jackson module would use the given Objectmapper feature. 
So that lazy loaded relations would shown only as id lists.

**Actual behavior**
Getting Lazyloading execptions.

**To Reproduce**
Steps to reproduce the behavior:
1. Clone https://github.com/cerias/quarkus-debug-jackson-objectmapper
2. setup database with one entity of TestA and TestB
3. Open /swagger-ui and trigger GET

**Configuration**
```properties
also in the example repo.
```

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: Linux huffelpuff 5.1.1-050101-generic #201905110631 SMP Sat May 11 06:33:50 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

 - Output of `java -version`: OpenJDK 64-Bit Server VM (build 11.0.4+11-post-Ubuntu-1ubuntu218.04.3, mixed mode, sharing)

 - GraalVM version (if different from Java): 19.0.2
 - Quarkus version or git rev: 0.25

It is also possible im not using the hibernate feature wrong. ;)

OK, so, what happens is that the serialisation of `TestA` which has a `@OneToMany` collection of `TestB` causes a lazy-loading exception:

```
Caused by: org.hibernate.LazyInitializationException: Unable to perform requested lazy initialization [wtf.heck.spqg.core.enviroment.entity.TestA.resources] - no session and settings disallow loading outside the Session
	at org.hibernate.bytecode.enhance.spi.interceptor.EnhancementHelper.throwLazyInitializationException(EnhancementHelper.java:196)
	at org.hibernate.bytecode.enhance.spi.interceptor.EnhancementHelper.performWork(EnhancementHelper.java:86)
	at org.hibernate.bytecode.enhance.spi.interceptor.LazyAttributeLoadingInterceptor.loadAttribute(LazyAttributeLoadingInterceptor.java:76)
	at org.hibernate.bytecode.enhance.spi.interceptor.LazyAttributeLoadingInterceptor.fetchAttribute(LazyAttributeLoadingInterceptor.java:72)
	at org.hibernate.bytecode.enhance.spi.interceptor.LazyAttributeLoadingInterceptor.handleRead(LazyAttributeLoadingInterceptor.java:53)
	at org.hibernate.bytecode.enhance.spi.interceptor.AbstractInterceptor.readObject(AbstractInterceptor.java:153)
	at wtf.heck.spqg.core.enviroment.entity.TestA.$$_hibernate_read_resources(TestA.java)
	at wtf.heck.spqg.core.enviroment.entity.TestA.getResources(TestA.java:38)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:688)
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:719)
```

Now, this exception is quite normal, _except_ that the jackson-hibernate module has some code to deal with lazy collections by opening sessions prior to reading them. Now, I don't know if that code would work, because it appears that code can't even get reached, because just reading the field/getter holding the collection will cause it to be loaded.

I believe that for non-enhanced entities you would probably be able to read the field/getter and get a `PersistentBag` which is not yet lazy-loaded and then Jackson-hibernate would try tricks with restoring a session to load it.

But in our case, just loading the field/getter triggers lazy-loading (even though in the end we do get a `PersistentBag`, but it's loaded) so Jackson-hibernate can't even look at its type to determine if it should load a session or not.

@Sanne is that a known behaviour for enhanced lazy collections?

Note that even if we were able to get the unloaded lazy collection it's really not clear if Jackson-hibernate would actually succeed in restoring a session.

Now, there's also code in Jackson-Hibernate that's dealing with `HibernateProxy` which we could rewrite to deal with `PersistentAttributeInterceptable` (as I see in `PersistenceUtilHelper.isLoaded` that it appears to be the type of bytecode-enhanced lazy proxies) but I haven't met such an instance in this test yet. I could expand the test to cause such an instance to be loaded by extending the model, but that does not appear to be the current issue's cause, even though future users are bound to run into it.

So, to recap, several issues:

- Calling `TestA.getResources()` (the lazy collection) causes the lazy collection to be loaded, where Jackson-Hibernate expects it to not be loaded. This may be different in non-enhanced entities.
- Jackson-Hibernate has code to restore sessions for lazy-loading but I've no idea if it would even work in Quarkus.
- Jackson-Hibernate has specific code to deal with `HibernateProxy` but not bytecode-enhanced proxies.

We could (well, really it would be better if someone from Hibernate would) participate upstream to Jackson-Hibernate to deal with these things in a more systematic way, but that appears to be a non-short-term task.

Not sure we can do something to make this work in time for 1.0, and while it's not ideal, there's a clear workaround which is to serialise using separate data classes. I would not call this a blocker, because there's a clear and complete workaround (even if verbose/inconvenient).

--- a/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/applicationfieldaccess/GetterAccessAssociationsTest.java
+++ b/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/applicationfieldaccess/GetterAccessAssociationsTest.java
@@ -1,4 +1,4 @@
-package io.quarkus.hibernate.orm.publicfields;
+package io.quarkus.hibernate.orm.applicationfieldaccess;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
@@ -30,10 +30,9 @@ import org.junit.jupiter.api.extension.RegisterExtension;
 import io.quarkus.test.QuarkusUnitTest;
 
 /**
- * Checks that public field access is correctly replaced with getter/setter calls,
- * regardless of the field type.
+ * Checks that access to fields through getters by the application works correctly for all association types.
  */
-public class PublicFieldAccessAssociationsTest {
+public class GetterAccessAssociationsTest {
 
     private static final String CONTAINED_VALUE = "someValue";
 
@@ -41,7 +40,7 @@ public class PublicFieldAccessAssociationsTest {
     static QuarkusUnitTest runner = new QuarkusUnitTest()
             .withApplicationRoot((jar) -> jar
                     .addClass(ContainingEntity.class)
-                    .addClass(FieldAccessEnhancedDelegate.class))
+                    .addClass(AccessDelegate.class))
             .withConfigurationResource("application-fetch-max-depth-zero.properties");
 
     @Inject
@@ -51,22 +50,22 @@ public class PublicFieldAccessAssociationsTest {
     UserTransaction transaction;
 
     @Test
-    public void testFieldAccess()
+    public void testGetterAccess()
             throws SystemException, NotSupportedException, HeuristicRollbackException, HeuristicMixedException,
             RollbackException {
         // Ideally we'd write a @ParameterizedTest and pass the delegates as parameters,
         // but we cannot do that due to JUnit using a different classloader than the test.
-        for (FieldAccessEnhancedDelegate delegate : FieldAccessEnhancedDelegate.values()) {
-            doTestFieldAccess(delegate);
+        for (AccessDelegate delegate : AccessDelegate.values()) {
+            doTestGetterAccess(delegate);
         }
     }
 
-    private void doTestFieldAccess(FieldAccessEnhancedDelegate delegate)
+    private void doTestGetterAccess(AccessDelegate delegate)
             throws SystemException, NotSupportedException, HeuristicRollbackException, HeuristicMixedException,
             RollbackException {
         ContainingEntity entity = new ContainingEntity();
         ContainedEntity containedEntity = new ContainedEntity();
-        containedEntity.value = CONTAINED_VALUE;
+        containedEntity.setValue(CONTAINED_VALUE);
 
         transaction.begin();
         em.persist(entity);
@@ -74,8 +73,8 @@ public class PublicFieldAccessAssociationsTest {
         transaction.commit();
 
         transaction.begin();
-        entity = em.getReference(ContainingEntity.class, entity.id);
-        containedEntity = em.getReference(ContainedEntity.class, containedEntity.id);
+        entity = em.getReference(ContainingEntity.class, entity.getId());
+        containedEntity = em.getReference(ContainedEntity.class, containedEntity.getId());
         // Initially the assertion doesn't pass: the value was not set yet
         AssertionError expected = null;
         try {
@@ -89,8 +88,8 @@ public class PublicFieldAccessAssociationsTest {
         transaction.rollback();
 
         transaction.begin();
-        entity = em.getReference(ContainingEntity.class, entity.id);
-        containedEntity = em.getReference(ContainedEntity.class, containedEntity.id);
+        entity = em.getReference(ContainingEntity.class, entity.getId());
+        containedEntity = em.getReference(ContainedEntity.class, containedEntity.getId());
         // Since field access is replaced with accessor calls,
         // we expect this change to be detected by dirty tracking and persisted.
         delegate.setValue(entity, containedEntity);
@@ -98,8 +97,8 @@ public class PublicFieldAccessAssociationsTest {
 
         // Test getReference()
         transaction.begin();
-        entity = em.getReference(ContainingEntity.class, entity.id);
-        containedEntity = em.getReference(ContainedEntity.class, containedEntity.id);
+        entity = em.getReference(ContainingEntity.class, entity.getId());
+        containedEntity = em.getReference(ContainedEntity.class, containedEntity.getId());
         // We're working on an uninitialized proxy.
         assertThat(entity).returns(false, Hibernate::isInitialized);
         // The above should have persisted a value that passes the assertion.
@@ -110,8 +109,8 @@ public class PublicFieldAccessAssociationsTest {
 
         // Test find()
         transaction.begin();
-        entity = em.find(ContainingEntity.class, entity.id);
-        containedEntity = em.find(ContainedEntity.class, containedEntity.id);
+        entity = em.find(ContainingEntity.class, entity.getId());
+        containedEntity = em.find(ContainedEntity.class, containedEntity.getId());
         // We're working on an actual entity instance (not a proxy).
         assertThat(entity).returns(true, Hibernate::isInitialized);
         // The above should have persisted a value that passes the assertion.
@@ -127,28 +126,91 @@ public class PublicFieldAccessAssociationsTest {
         public long id;
 
         @OneToOne
-        public ContainedEntity oneToOne;
+        private ContainedEntity oneToOne;
 
         @ManyToOne
-        public ContainedEntity manyToOne;
+        private ContainedEntity manyToOne;
 
         @OneToMany
         @JoinTable(name = "containing_oneToMany")
-        public List<ContainedEntity> oneToMany = new ArrayList<>();
+        private List<ContainedEntity> oneToMany = new ArrayList<>();
 
         @ManyToMany
         @JoinTable(name = "containing_manyToMany")
-        public List<ContainedEntity> manyToMany = new ArrayList<>();
+        private List<ContainedEntity> manyToMany = new ArrayList<>();
 
         @OneToOne(mappedBy = "oneToOne")
-        public ContainedEntity oneToOneMappedBy;
+        private ContainedEntity oneToOneMappedBy;
 
         @OneToMany(mappedBy = "manyToOne")
-        public List<ContainedEntity> oneToManyMappedBy = new ArrayList<>();
+        private List<ContainedEntity> oneToManyMappedBy = new ArrayList<>();
 
         @ManyToMany(mappedBy = "manyToMany")
-        public List<ContainedEntity> manyToManyMappedBy = new ArrayList<>();
+        private List<ContainedEntity> manyToManyMappedBy = new ArrayList<>();
 
+        public long getId() {
+            return id;
+        }
+
+        public void setId(long id) {
+            this.id = id;
+        }
+
+        public ContainedEntity getOneToOne() {
+            return oneToOne;
+        }
+
+        public void setOneToOne(ContainedEntity oneToOne) {
+            this.oneToOne = oneToOne;
+        }
+
+        public ContainedEntity getManyToOne() {
+            return manyToOne;
+        }
+
+        public void setManyToOne(ContainedEntity manyToOne) {
+            this.manyToOne = manyToOne;
+        }
+
+        public List<ContainedEntity> getOneToMany() {
+            return oneToMany;
+        }
+
+        public void setOneToMany(List<ContainedEntity> oneToMany) {
+            this.oneToMany = oneToMany;
+        }
+
+        public List<ContainedEntity> getManyToMany() {
+            return manyToMany;
+        }
+
+        public void setManyToMany(List<ContainedEntity> manyToMany) {
+            this.manyToMany = manyToMany;
+        }
+
+        public ContainedEntity getOneToOneMappedBy() {
+            return oneToOneMappedBy;
+        }
+
+        public void setOneToOneMappedBy(ContainedEntity oneToOneMappedBy) {
+            this.oneToOneMappedBy = oneToOneMappedBy;
+        }
+
+        public List<ContainedEntity> getOneToManyMappedBy() {
+            return oneToManyMappedBy;
+        }
+
+        public void setOneToManyMappedBy(List<ContainedEntity> oneToManyMappedBy) {
+            this.oneToManyMappedBy = oneToManyMappedBy;
+        }
+
+        public List<ContainedEntity> getManyToManyMappedBy() {
+            return manyToManyMappedBy;
+        }
+
+        public void setManyToManyMappedBy(List<ContainedEntity> manyToManyMappedBy) {
+            this.manyToManyMappedBy = manyToManyMappedBy;
+        }
     }
 
     @Entity
@@ -156,122 +218,161 @@ public class PublicFieldAccessAssociationsTest {
 
         @Id
         @GeneratedValue
-        public long id;
+        private long id;
 
         @Column(name = "value_")
-        public String value;
+        private String value;
 
         @OneToOne
-        public ContainingEntity oneToOne;
+        private ContainingEntity oneToOne;
 
         @ManyToOne
-        public ContainingEntity manyToOne;
+        private ContainingEntity manyToOne;
 
         @ManyToMany
         @JoinTable(name = "containing_manyToMany_mappedBy")
-        public List<ContainingEntity> manyToMany = new ArrayList<>();
+        private List<ContainingEntity> manyToMany = new ArrayList<>();
 
+        public long getId() {
+            return id;
+        }
+
+        public void setId(long id) {
+            this.id = id;
+        }
+
+        public String getValue() {
+            return value;
+        }
+
+        public void setValue(String value) {
+            this.value = value;
+        }
+
+        public ContainingEntity getOneToOne() {
+            return oneToOne;
+        }
+
+        public void setOneToOne(ContainingEntity oneToOne) {
+            this.oneToOne = oneToOne;
+        }
+
+        public ContainingEntity getManyToOne() {
+            return manyToOne;
+        }
+
+        public void setManyToOne(ContainingEntity manyToOne) {
+            this.manyToOne = manyToOne;
+        }
+
+        public List<ContainingEntity> getManyToMany() {
+            return manyToMany;
+        }
+
+        public void setManyToMany(List<ContainingEntity> manyToMany) {
+            this.manyToMany = manyToMany;
+        }
     }
 
-    private enum FieldAccessEnhancedDelegate {
+    private enum AccessDelegate {
 
         ONE_TO_ONE {
             @Override
             public void setValue(ContainingEntity entity, ContainedEntity containedEntity) {
-                entity.oneToOne = containedEntity;
+                entity.setOneToOne(containedEntity);
             }
 
             @Override
             public void assertValueAndLaziness(ContainingEntity entity, ContainedEntity containedEntity) {
                 // No expectations regarding laziness on ToOne associations
-                assertThat(entity.oneToOne).isEqualTo(containedEntity);
-                consumeValue(entity.oneToOne);
+                assertThat(entity.getOneToOne()).isEqualTo(containedEntity);
+                consumeValue(entity.getOneToOne());
             }
         },
         MANY_TO_ONE {
             @Override
             public void setValue(ContainingEntity entity, ContainedEntity containedEntity) {
-                entity.manyToOne = containedEntity;
+                entity.setManyToOne(containedEntity);
             }
 
             @Override
             public void assertValueAndLaziness(ContainingEntity entity, ContainedEntity containedEntity) {
                 // No expectations regarding laziness on ToOne associations
-                assertThat(entity.manyToOne).isEqualTo(containedEntity);
-                consumeValue(entity.manyToOne);
+                assertThat(entity.getManyToOne()).isEqualTo(containedEntity);
+                consumeValue(entity.getManyToOne());
             }
         },
         ONE_TO_MANY {
             @Override
             public void setValue(ContainingEntity entity, ContainedEntity containedEntity) {
-                entity.oneToMany.add(containedEntity);
+                entity.getOneToMany().add(containedEntity);
             }
 
             @Override
             public void assertValueAndLaziness(ContainingEntity entity, ContainedEntity containedEntity) {
-                assertThat((Object) entity.oneToMany).returns(false, Hibernate::isInitialized);
-                assertThat(entity.oneToMany).containsExactly(containedEntity);
-                assertThat((Object) entity.oneToMany).returns(true, Hibernate::isInitialized);
+                assertThat((Object) entity.getOneToMany()).returns(false, Hibernate::isInitialized);
+                assertThat(entity.getOneToMany()).containsExactly(containedEntity);
+                assertThat((Object) entity.getOneToMany()).returns(true, Hibernate::isInitialized);
             }
         },
         MANY_TO_MANY {
             @Override
             public void setValue(ContainingEntity entity, ContainedEntity containedEntity) {
-                entity.manyToMany.add(containedEntity);
+                entity.getManyToMany().add(containedEntity);
             }
 
             @Override
             public void assertValueAndLaziness(ContainingEntity entity, ContainedEntity containedEntity) {
-                assertThat((Object) entity.manyToMany).returns(false, Hibernate::isInitialized);
-                assertThat(entity.manyToMany).containsExactly(containedEntity);
-                assertThat((Object) entity.manyToMany).returns(true, Hibernate::isInitialized);
+                assertThat((Object) entity.getManyToMany()).returns(false, Hibernate::isInitialized);
+                assertThat(entity.getManyToMany()).containsExactly(containedEntity);
+                assertThat((Object) entity.getManyToMany()).returns(true, Hibernate::isInitialized);
             }
         },
         ONE_TO_ONE_MAPPED_BY {
             @Override
             public void setValue(ContainingEntity entity, ContainedEntity containedEntity) {
-                entity.oneToOneMappedBy = containedEntity;
-                containedEntity.oneToOne = entity;
+                entity.setOneToOneMappedBy(containedEntity);
+                containedEntity.setOneToOne(entity);
             }
 
             @Override
             public void assertValueAndLaziness(ContainingEntity entity, ContainedEntity containedEntity) {
                 // No expectations regarding laziness on ToOne associations
-                assertThat(entity.oneToOneMappedBy).isEqualTo(containedEntity);
-                consumeValue(entity.oneToOneMappedBy);
+                assertThat(entity.getOneToOneMappedBy()).isEqualTo(containedEntity);
+                consumeValue(entity.getOneToOneMappedBy());
             }
         },
         ONE_TO_MANY_MAPPED_BY {
             @Override
             public void setValue(ContainingEntity entity, ContainedEntity containedEntity) {
-                entity.oneToManyMappedBy.add(containedEntity);
-                containedEntity.manyToOne = entity;
+                entity.getOneToManyMappedBy().add(containedEntity);
+                containedEntity.setManyToOne(entity);
             }
 
             @Override
             public void assertValueAndLaziness(ContainingEntity entity, ContainedEntity containedEntity) {
-                assertThat((Object) entity.oneToManyMappedBy).returns(false, Hibernate::isInitialized);
-                assertThat(entity.oneToManyMappedBy).containsExactly(containedEntity);
-                assertThat((Object) entity.oneToManyMappedBy).returns(true, Hibernate::isInitialized);
+                assertThat((Object) entity.getOneToManyMappedBy()).returns(false, Hibernate::isInitialized);
+                assertThat(entity.getOneToManyMappedBy()).containsExactly(containedEntity);
+                assertThat((Object) entity.getOneToManyMappedBy()).returns(true, Hibernate::isInitialized);
             }
         },
         MANY_TO_MANY_MAPPED_BY {
             @Override
             public void setValue(ContainingEntity entity, ContainedEntity containedEntity) {
-                entity.manyToManyMappedBy.add(containedEntity);
-                containedEntity.manyToMany.add(entity);
+                entity.getManyToManyMappedBy().add(containedEntity);
+                containedEntity.getManyToMany().add(entity);
             }
 
             @Override
             public void assertValueAndLaziness(ContainingEntity entity, ContainedEntity containedEntity) {
-                assertThat((Object) entity.manyToManyMappedBy).returns(false, Hibernate::isInitialized);
-                assertThat(entity.manyToManyMappedBy).containsExactly(containedEntity);
-                assertThat((Object) entity.manyToManyMappedBy).returns(true, Hibernate::isInitialized);
+                assertThat((Object) entity.getManyToManyMappedBy()).returns(false, Hibernate::isInitialized);
+                assertThat(entity.getManyToManyMappedBy()).containsExactly(containedEntity);
+                assertThat((Object) entity.getManyToManyMappedBy()).returns(true, Hibernate::isInitialized);
             }
         };
 
         protected void consumeValue(ContainedEntity entity) {
-            assertThat(entity.value).isEqualTo(CONTAINED_VALUE);
+            assertThat(entity.getValue()).isEqualTo(CONTAINED_VALUE);
         }
 
         public abstract void setValue(ContainingEntity entity, ContainedEntity containedEntity);

--- a/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/applicationfieldaccess/GetterAccessAssociationsTest.java
+++ b/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/applicationfieldaccess/GetterAccessAssociationsTest.java
@@ -1,6 +1,8 @@
 package io.quarkus.hibernate.orm.applicationfieldaccess;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatCode;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
 import java.util.ArrayList;
 import java.util.List;
@@ -24,6 +26,7 @@ import javax.transaction.SystemException;
 import javax.transaction.UserTransaction;
 
 import org.hibernate.Hibernate;
+import org.hibernate.LazyInitializationException;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
@@ -116,6 +119,17 @@ public class GetterAccessAssociationsTest {
         // The above should have persisted a value that passes the assertion.
         delegate.assertValueAndLaziness(entity, containedEntity);
         transaction.commit();
+
+        // Test find() + access outside of session
+        transaction.begin();
+        entity = em.find(ContainingEntity.class, entity.getId());
+        containedEntity = em.find(ContainedEntity.class, containedEntity.getId());
+        // We're working on an actual entity instance (not a proxy).
+        assertThat(entity).returns(true, Hibernate::isInitialized);
+        transaction.commit();
+        // Access to the property out of session should pass certain assertions
+        // (which are different depending on the type of association).
+        delegate.assertAccessOutOfSession(entity, containedEntity);
     }
 
     @Entity
@@ -288,6 +302,13 @@ public class GetterAccessAssociationsTest {
                 assertThat(entity.getOneToOne()).isEqualTo(containedEntity);
                 consumeValue(entity.getOneToOne());
             }
+
+            @Override
+            public void assertAccessOutOfSession(ContainingEntity entity, ContainedEntity containedEntity) {
+                // No expectations regarding laziness on ToOne associations
+                assertThat(entity.getOneToOne()).isEqualTo(containedEntity);
+                consumeValue(entity.getOneToOne());
+            }
         },
         MANY_TO_ONE {
             @Override
@@ -297,6 +318,13 @@ public class GetterAccessAssociationsTest {
 
             @Override
             public void assertValueAndLaziness(ContainingEntity entity, ContainedEntity containedEntity) {
+                // No expectations regarding laziness on ToOne associations
+                assertThat(entity.getManyToOne()).isEqualTo(containedEntity);
+                consumeValue(entity.getManyToOne());
+            }
+
+            @Override
+            public void assertAccessOutOfSession(ContainingEntity entity, ContainedEntity containedEntity) {
                 // No expectations regarding laziness on ToOne associations
                 assertThat(entity.getManyToOne()).isEqualTo(containedEntity);
                 consumeValue(entity.getManyToOne());
@@ -314,6 +342,17 @@ public class GetterAccessAssociationsTest {
                 assertThat(entity.getOneToMany()).containsExactly(containedEntity);
                 assertThat((Object) entity.getOneToMany()).returns(true, Hibernate::isInitialized);
             }
+
+            @Override
+            public void assertAccessOutOfSession(ContainingEntity entity, ContainedEntity containedEntity) {
+                // We expect to be able to call the getter outside of the session on an initialized entity.
+                // https://github.com/quarkusio/quarkus/discussions/27657
+                assertThatCode(() -> entity.getOneToMany()).doesNotThrowAnyException();
+                // But of course, the collection is not initialized and accessing the content won't work.
+                var collection = entity.getOneToMany();
+                assertThat((Object) collection).returns(false, Hibernate::isInitialized);
+                assertThatThrownBy(() -> collection.size()).isInstanceOf(LazyInitializationException.class);
+            }
         },
         MANY_TO_MANY {
             @Override
@@ -327,6 +366,17 @@ public class GetterAccessAssociationsTest {
                 assertThat(entity.getManyToMany()).containsExactly(containedEntity);
                 assertThat((Object) entity.getManyToMany()).returns(true, Hibernate::isInitialized);
             }
+
+            @Override
+            public void assertAccessOutOfSession(ContainingEntity entity, ContainedEntity containedEntity) {
+                // We expect to be able to call the getter outside of the session on an initialized entity.
+                // https://github.com/quarkusio/quarkus/discussions/27657
+                assertThatCode(() -> entity.getManyToMany()).doesNotThrowAnyException();
+                // But of course, the collection is not initialized and accessing the content won't work.
+                var collection = entity.getManyToMany();
+                assertThat((Object) collection).returns(false, Hibernate::isInitialized);
+                assertThatThrownBy(() -> collection.size()).isInstanceOf(LazyInitializationException.class);
+            }
         },
         ONE_TO_ONE_MAPPED_BY {
             @Override
@@ -337,6 +387,13 @@ public class GetterAccessAssociationsTest {
 
             @Override
             public void assertValueAndLaziness(ContainingEntity entity, ContainedEntity containedEntity) {
+                // No expectations regarding laziness on ToOne associations
+                assertThat(entity.getOneToOneMappedBy()).isEqualTo(containedEntity);
+                consumeValue(entity.getOneToOneMappedBy());
+            }
+
+            @Override
+            public void assertAccessOutOfSession(ContainingEntity entity, ContainedEntity containedEntity) {
                 // No expectations regarding laziness on ToOne associations
                 assertThat(entity.getOneToOneMappedBy()).isEqualTo(containedEntity);
                 consumeValue(entity.getOneToOneMappedBy());
@@ -355,6 +412,17 @@ public class GetterAccessAssociationsTest {
                 assertThat(entity.getOneToManyMappedBy()).containsExactly(containedEntity);
                 assertThat((Object) entity.getOneToManyMappedBy()).returns(true, Hibernate::isInitialized);
             }
+
+            @Override
+            public void assertAccessOutOfSession(ContainingEntity entity, ContainedEntity containedEntity) {
+                // We expect to be able to call the getter outside of the session on an initialized entity.
+                // https://github.com/quarkusio/quarkus/discussions/27657
+                assertThatCode(() -> entity.getOneToManyMappedBy()).doesNotThrowAnyException();
+                // But of course, the collection is not initialized and accessing the content won't work.
+                var collection = entity.getOneToManyMappedBy();
+                assertThat((Object) collection).returns(false, Hibernate::isInitialized);
+                assertThatThrownBy(() -> collection.size()).isInstanceOf(LazyInitializationException.class);
+            }
         },
         MANY_TO_MANY_MAPPED_BY {
             @Override
@@ -369,6 +437,17 @@ public class GetterAccessAssociationsTest {
                 assertThat(entity.getManyToManyMappedBy()).containsExactly(containedEntity);
                 assertThat((Object) entity.getManyToManyMappedBy()).returns(true, Hibernate::isInitialized);
             }
+
+            @Override
+            public void assertAccessOutOfSession(ContainingEntity entity, ContainedEntity containedEntity) {
+                // We expect to be able to call the getter outside of the session on an initialized entity.
+                // https://github.com/quarkusio/quarkus/discussions/27657
+                assertThatCode(() -> entity.getManyToManyMappedBy()).doesNotThrowAnyException();
+                // But of course, the collection is not initialized and accessing the content won't work.
+                var collection = entity.getManyToManyMappedBy();
+                assertThat((Object) collection).returns(false, Hibernate::isInitialized);
+                assertThatThrownBy(() -> collection.size()).isInstanceOf(LazyInitializationException.class);
+            }
         };
 
         protected void consumeValue(ContainedEntity entity) {
@@ -378,6 +457,8 @@ public class GetterAccessAssociationsTest {
         public abstract void setValue(ContainingEntity entity, ContainedEntity containedEntity);
 
         public abstract void assertValueAndLaziness(ContainingEntity entity, ContainedEntity containedEntity);
+
+        public abstract void assertAccessOutOfSession(ContainingEntity entity, ContainedEntity containedEntity);
 
     }
 }

--- a/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/applicationfieldaccess/PublicFieldAccessAssociationsTest.java
+++ b/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/applicationfieldaccess/PublicFieldAccessAssociationsTest.java
@@ -1,0 +1,282 @@
+package io.quarkus.hibernate.orm.applicationfieldaccess;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.EntityManager;
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+import javax.persistence.JoinTable;
+import javax.persistence.ManyToMany;
+import javax.persistence.ManyToOne;
+import javax.persistence.OneToMany;
+import javax.persistence.OneToOne;
+import javax.transaction.HeuristicMixedException;
+import javax.transaction.HeuristicRollbackException;
+import javax.transaction.NotSupportedException;
+import javax.transaction.RollbackException;
+import javax.transaction.SystemException;
+import javax.transaction.UserTransaction;
+
+import org.hibernate.Hibernate;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+/**
+ * Checks that access to public fields by the application is correctly replaced with getter/setter calls
+ * and works correctly for all association types.
+ */
+public class PublicFieldAccessAssociationsTest {
+
+    private static final String CONTAINED_VALUE = "someValue";
+
+    @RegisterExtension
+    static QuarkusUnitTest runner = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClass(ContainingEntity.class)
+                    .addClass(FieldAccessEnhancedDelegate.class))
+            .withConfigurationResource("application-fetch-max-depth-zero.properties");
+
+    @Inject
+    EntityManager em;
+
+    @Inject
+    UserTransaction transaction;
+
+    @Test
+    public void testFieldAccess()
+            throws SystemException, NotSupportedException, HeuristicRollbackException, HeuristicMixedException,
+            RollbackException {
+        // Ideally we'd write a @ParameterizedTest and pass the delegates as parameters,
+        // but we cannot do that due to JUnit using a different classloader than the test.
+        for (FieldAccessEnhancedDelegate delegate : FieldAccessEnhancedDelegate.values()) {
+            doTestFieldAccess(delegate);
+        }
+    }
+
+    private void doTestFieldAccess(FieldAccessEnhancedDelegate delegate)
+            throws SystemException, NotSupportedException, HeuristicRollbackException, HeuristicMixedException,
+            RollbackException {
+        ContainingEntity entity = new ContainingEntity();
+        ContainedEntity containedEntity = new ContainedEntity();
+        containedEntity.value = CONTAINED_VALUE;
+
+        transaction.begin();
+        em.persist(entity);
+        em.persist(containedEntity);
+        transaction.commit();
+
+        transaction.begin();
+        entity = em.getReference(ContainingEntity.class, entity.id);
+        containedEntity = em.getReference(ContainedEntity.class, containedEntity.id);
+        // Initially the assertion doesn't pass: the value was not set yet
+        AssertionError expected = null;
+        try {
+            delegate.assertValueAndLaziness(entity, containedEntity);
+        } catch (AssertionError e) {
+            expected = e;
+        }
+        if (expected == null) {
+            throw new IllegalStateException("This test is buggy: assertions should not pass at this point.");
+        }
+        transaction.rollback();
+
+        transaction.begin();
+        entity = em.getReference(ContainingEntity.class, entity.id);
+        containedEntity = em.getReference(ContainedEntity.class, containedEntity.id);
+        // Since field access is replaced with accessor calls,
+        // we expect this change to be detected by dirty tracking and persisted.
+        delegate.setValue(entity, containedEntity);
+        transaction.commit();
+
+        // Test getReference()
+        transaction.begin();
+        entity = em.getReference(ContainingEntity.class, entity.id);
+        containedEntity = em.getReference(ContainedEntity.class, containedEntity.id);
+        // We're working on an uninitialized proxy.
+        assertThat(entity).returns(false, Hibernate::isInitialized);
+        // The above should have persisted a value that passes the assertion.
+        delegate.assertValueAndLaziness(entity, containedEntity);
+        // Accessing the value should trigger initialization of the proxy.
+        assertThat(entity).returns(true, Hibernate::isInitialized);
+        transaction.commit();
+
+        // Test find()
+        transaction.begin();
+        entity = em.find(ContainingEntity.class, entity.id);
+        containedEntity = em.find(ContainedEntity.class, containedEntity.id);
+        // We're working on an actual entity instance (not a proxy).
+        assertThat(entity).returns(true, Hibernate::isInitialized);
+        // The above should have persisted a value that passes the assertion.
+        delegate.assertValueAndLaziness(entity, containedEntity);
+        transaction.commit();
+    }
+
+    @Entity
+    public static class ContainingEntity {
+
+        @Id
+        @GeneratedValue
+        public long id;
+
+        @OneToOne
+        public ContainedEntity oneToOne;
+
+        @ManyToOne
+        public ContainedEntity manyToOne;
+
+        @OneToMany
+        @JoinTable(name = "containing_oneToMany")
+        public List<ContainedEntity> oneToMany = new ArrayList<>();
+
+        @ManyToMany
+        @JoinTable(name = "containing_manyToMany")
+        public List<ContainedEntity> manyToMany = new ArrayList<>();
+
+        @OneToOne(mappedBy = "oneToOne")
+        public ContainedEntity oneToOneMappedBy;
+
+        @OneToMany(mappedBy = "manyToOne")
+        public List<ContainedEntity> oneToManyMappedBy = new ArrayList<>();
+
+        @ManyToMany(mappedBy = "manyToMany")
+        public List<ContainedEntity> manyToManyMappedBy = new ArrayList<>();
+
+    }
+
+    @Entity
+    public static class ContainedEntity {
+
+        @Id
+        @GeneratedValue
+        public long id;
+
+        @Column(name = "value_")
+        public String value;
+
+        @OneToOne
+        public ContainingEntity oneToOne;
+
+        @ManyToOne
+        public ContainingEntity manyToOne;
+
+        @ManyToMany
+        @JoinTable(name = "containing_manyToMany_mappedBy")
+        public List<ContainingEntity> manyToMany = new ArrayList<>();
+
+    }
+
+    private enum FieldAccessEnhancedDelegate {
+
+        ONE_TO_ONE {
+            @Override
+            public void setValue(ContainingEntity entity, ContainedEntity containedEntity) {
+                entity.oneToOne = containedEntity;
+            }
+
+            @Override
+            public void assertValueAndLaziness(ContainingEntity entity, ContainedEntity containedEntity) {
+                // No expectations regarding laziness on ToOne associations
+                assertThat(entity.oneToOne).isEqualTo(containedEntity);
+                consumeValue(entity.oneToOne);
+            }
+        },
+        MANY_TO_ONE {
+            @Override
+            public void setValue(ContainingEntity entity, ContainedEntity containedEntity) {
+                entity.manyToOne = containedEntity;
+            }
+
+            @Override
+            public void assertValueAndLaziness(ContainingEntity entity, ContainedEntity containedEntity) {
+                // No expectations regarding laziness on ToOne associations
+                assertThat(entity.manyToOne).isEqualTo(containedEntity);
+                consumeValue(entity.manyToOne);
+            }
+        },
+        ONE_TO_MANY {
+            @Override
+            public void setValue(ContainingEntity entity, ContainedEntity containedEntity) {
+                entity.oneToMany.add(containedEntity);
+            }
+
+            @Override
+            public void assertValueAndLaziness(ContainingEntity entity, ContainedEntity containedEntity) {
+                assertThat((Object) entity.oneToMany).returns(false, Hibernate::isInitialized);
+                assertThat(entity.oneToMany).containsExactly(containedEntity);
+                assertThat((Object) entity.oneToMany).returns(true, Hibernate::isInitialized);
+            }
+        },
+        MANY_TO_MANY {
+            @Override
+            public void setValue(ContainingEntity entity, ContainedEntity containedEntity) {
+                entity.manyToMany.add(containedEntity);
+            }
+
+            @Override
+            public void assertValueAndLaziness(ContainingEntity entity, ContainedEntity containedEntity) {
+                assertThat((Object) entity.manyToMany).returns(false, Hibernate::isInitialized);
+                assertThat(entity.manyToMany).containsExactly(containedEntity);
+                assertThat((Object) entity.manyToMany).returns(true, Hibernate::isInitialized);
+            }
+        },
+        ONE_TO_ONE_MAPPED_BY {
+            @Override
+            public void setValue(ContainingEntity entity, ContainedEntity containedEntity) {
+                entity.oneToOneMappedBy = containedEntity;
+                containedEntity.oneToOne = entity;
+            }
+
+            @Override
+            public void assertValueAndLaziness(ContainingEntity entity, ContainedEntity containedEntity) {
+                // No expectations regarding laziness on ToOne associations
+                assertThat(entity.oneToOneMappedBy).isEqualTo(containedEntity);
+                consumeValue(entity.oneToOneMappedBy);
+            }
+        },
+        ONE_TO_MANY_MAPPED_BY {
+            @Override
+            public void setValue(ContainingEntity entity, ContainedEntity containedEntity) {
+                entity.oneToManyMappedBy.add(containedEntity);
+                containedEntity.manyToOne = entity;
+            }
+
+            @Override
+            public void assertValueAndLaziness(ContainingEntity entity, ContainedEntity containedEntity) {
+                assertThat((Object) entity.oneToManyMappedBy).returns(false, Hibernate::isInitialized);
+                assertThat(entity.oneToManyMappedBy).containsExactly(containedEntity);
+                assertThat((Object) entity.oneToManyMappedBy).returns(true, Hibernate::isInitialized);
+            }
+        },
+        MANY_TO_MANY_MAPPED_BY {
+            @Override
+            public void setValue(ContainingEntity entity, ContainedEntity containedEntity) {
+                entity.manyToManyMappedBy.add(containedEntity);
+                containedEntity.manyToMany.add(entity);
+            }
+
+            @Override
+            public void assertValueAndLaziness(ContainingEntity entity, ContainedEntity containedEntity) {
+                assertThat((Object) entity.manyToManyMappedBy).returns(false, Hibernate::isInitialized);
+                assertThat(entity.manyToManyMappedBy).containsExactly(containedEntity);
+                assertThat((Object) entity.manyToManyMappedBy).returns(true, Hibernate::isInitialized);
+            }
+        };
+
+        protected void consumeValue(ContainedEntity entity) {
+            assertThat(entity.value).isEqualTo(CONTAINED_VALUE);
+        }
+
+        public abstract void setValue(ContainingEntity entity, ContainedEntity containedEntity);
+
+        public abstract void assertValueAndLaziness(ContainingEntity entity, ContainedEntity containedEntity);
+
+    }
+}

--- a/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/applicationfieldaccess/PublicFieldAccessFieldTypesTest.java
+++ b/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/applicationfieldaccess/PublicFieldAccessFieldTypesTest.java
@@ -1,4 +1,4 @@
-package io.quarkus.hibernate.orm.publicfields;
+package io.quarkus.hibernate.orm.applicationfieldaccess;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
@@ -23,8 +23,8 @@ import org.junit.jupiter.api.extension.RegisterExtension;
 import io.quarkus.test.QuarkusUnitTest;
 
 /**
- * Checks that public field access is correctly replaced with getter/setter calls,
- * regardless of the field type.
+ * Checks that access to public fields by the application is correctly replaced with getter/setter calls
+ * and works correctly for all field types.
  */
 public class PublicFieldAccessFieldTypesTest {
 

--- a/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/applicationfieldaccess/PublicFieldAccessFinalFieldTest.java
+++ b/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/applicationfieldaccess/PublicFieldAccessFinalFieldTest.java
@@ -4,7 +4,7 @@
  * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
  * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
  */
-package io.quarkus.hibernate.orm.publicfields;
+package io.quarkus.hibernate.orm.applicationfieldaccess;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
@@ -30,7 +30,7 @@ import io.quarkus.hibernate.orm.TransactionTestUtils;
 import io.quarkus.test.QuarkusUnitTest;
 
 /**
- * Checks that public, final field access is correctly replaced with getter calls for reads,
+ * Checks that access to public, final fields by the application is correctly replaced with getter calls for reads,
  * but not replaced at all for writes (since writes to final fields can only occur from constructors).
  * <p>
  * See https://github.com/quarkusio/quarkus/issues/20186

--- a/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/applicationfieldaccess/PublicFieldAccessInheritanceTest.java
+++ b/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/applicationfieldaccess/PublicFieldAccessInheritanceTest.java
@@ -1,4 +1,4 @@
-package io.quarkus.hibernate.orm.publicfields;
+package io.quarkus.hibernate.orm.applicationfieldaccess;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
@@ -22,8 +22,8 @@ import org.junit.jupiter.api.extension.RegisterExtension;
 import io.quarkus.test.QuarkusUnitTest;
 
 /**
- * Checks that public field access is correctly replaced with getter/setter calls,
- * regardless of where the field is declared in the class hierarchy.
+ * Checks that access to public fields by the application is correctly replaced with getter/setter calls
+ * and works correctly regardless of where the field is declared in the class hierarchy.
  */
 public class PublicFieldAccessInheritanceTest {
 

--- a/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/applicationfieldaccess/PublicFieldWithProxyAndLazyLoadingAndInheritanceTest.java
+++ b/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/applicationfieldaccess/PublicFieldWithProxyAndLazyLoadingAndInheritanceTest.java
@@ -4,7 +4,7 @@
  * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
  * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
  */
-package io.quarkus.hibernate.orm.publicfields;
+package io.quarkus.hibernate.orm.applicationfieldaccess;
 
 import static io.quarkus.hibernate.orm.TransactionTestUtils.inTransaction;
 import static org.assertj.core.api.Assertions.assertThat;

--- a/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/batch/BatchFetchSizeTest.java
+++ b/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/batch/BatchFetchSizeTest.java
@@ -45,20 +45,20 @@ public class BatchFetchSizeTest {
 
         transaction.begin();
         List<MainEntity> entities = session.createQuery("from MainEntity", MainEntity.class).list();
-        assertThat(entities).allSatisfy(e -> assertThat(Hibernate.isPropertyInitialized(e, "others"))
+        assertThat(entities).allSatisfy(e -> assertThat(Hibernate.isInitialized(e.others))
                 .as("'others' initialized for " + e).isFalse());
 
         MainEntity entity = entities.get(0);
         // Trigger initialization for the collection from one entity.
         entity.others.get(0);
-        assertThat(Hibernate.isPropertyInitialized(entity, "others"))
+        assertThat(Hibernate.isInitialized(entity.others))
                 .as("'others' initialized for " + entity).isTrue();
 
         // 20 entities were already in the session when the initialization above occurred,
         // so it should have triggered batch initialization of several 'others' collections.
         assertThat(entities).hasSize(20);
         assertThat(entities)
-                .filteredOn(e -> Hibernate.isPropertyInitialized(e, "others"))
+                .filteredOn(e -> Hibernate.isInitialized(e.others))
                 .hasSize(16); // Default batch fetch size is 16
         transaction.commit();
     }

--- a/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/FastBootEntityManagerFactoryBuilder.java
+++ b/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/FastBootEntityManagerFactoryBuilder.java
@@ -121,7 +121,6 @@ public class FastBootEntityManagerFactoryBuilder implements EntityManagerFactory
     }
 
     protected void populate(String persistenceUnitName, SessionFactoryOptionsBuilder options, StandardServiceRegistry ssr) {
-
         // will use user override value or default to false if not supplied to follow
         // JPA spec.
         final boolean jtaTransactionAccessEnabled = runtimeSettings.getBoolean(
@@ -157,6 +156,11 @@ public class FastBootEntityManagerFactoryBuilder implements EntityManagerFactory
         options.addSessionFactoryObservers(new ServiceRegistryCloser());
 
         options.applyEntityNotFoundDelegate(new JpaEntityNotFoundDelegate());
+
+        // This is necessary for Hibernate Reactive, see https://github.com/quarkusio/quarkus/issues/15814
+        // This is also necessary for Hibernate ORM if we want to prevent calls to getters on initialized entities
+        // outside of sessions from throwing exceptions, see https://github.com/quarkusio/quarkus/discussions/27657
+        options.enableCollectionInDefaultFetchGroup(true);
 
         if (this.validatorFactory != null) {
             options.applyValidatorFactory(validatorFactory);

